crates/sui-adapter/src/adapter.rs

```diff
@@ -992,136 +1009,63 @@ pub fn resolve_and_type_check(
-                                EntryArgumentErrorKind::InvalidObjectByMuteRef,
-                            ),
-                            error,
-                        ));
-                    }
-                    mutable_ref_objects.insert(idx as LocalIndex, id);
-                    &**inner_t
-                }
-                t @ SignatureToken::Struct(_)
-                | t @ SignatureToken::StructInstantiation(_, _)
-                | t @ SignatureToken::TypeParameter(_) => {
-                    match &object.owner {
-                        Owner::AddressOwner(_) | Owner::ObjectOwner(_) => (),
-                        Owner::Shared | Owner::Immutable => {
-                            return Err(ExecutionError::new_with_source(
-                                ExecutionErrorKind::entry_argument_error(
-                                    idx,
-                                    EntryArgumentErrorKind::InvalidObjectByValue,
-                                ),
-                                format!(
-                                    "Immutable and shared objects cannot be passed by-value, \
-                                    violation found in argument {}",
-                                    idx
-                                ),
-                            ));
-                        }
+                CallArg::Object(ObjectArg::ImmOrOwnedObject(ref_)) => serialize_object(
+                    InputObjectKind::ImmOrOwnedMoveObject(ref_),
+                    idx,
+                    param_type,
+                    objects,
+                    &mut object_data,
+                    &mut mutable_ref_objects,
+                    &mut by_value_objects,
+                    &mut object_type_map,
+                )?,
+                CallArg::Object(ObjectArg::SharedObject(id)) => serialize_object(
+                    InputObjectKind::SharedMoveObject(id),
+                    idx,
+                    param_type,
+                    objects,
+                    &mut object_data,
+                    &mut mutable_ref_objects,
+                    &mut by_value_objects,
+                    &mut object_type_map,
+                )?,
+                CallArg::ObjVec(vec) => {
+                    if vec.is_empty() {
+                        // bcs representation of the empty vector
+                        return Ok(vec![0]);
                     }
-                    by_value_objects.insert(id);
-                    t
-                }
-                t => {
-                    return Err(ExecutionError::new_with_source(
-                        ExecutionErrorKind::entry_argument_error(
+                    // write length of the vector as uleb128 as it is encoded in BCS and then append
+                    // all (already serialized) object content data
+                    let mut res = vec![];
+                    leb128::write::unsigned(&mut res, vec.len() as u64).unwrap();
```

Comment:
bit funky, but seems fine for now :)

Reply:
> bit funky, but seems fine for now :)

It is a little, isn't it? :-) Could not think of a better/simpler way, though...